{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tunable Ribosome Occupancy Model\n",
    "_Author: Raghav Chanchani_\n",
    "\n",
    "_For the Brian Zid laboratory at UC San Diego_\n",
    "\n",
    "_Date last modified: 10/4/2018_\n",
    "#### A discrete-time Markov chain to simulate ribosome distribution along a given gene using the assumptions listed below. There is no final absorption state. Ribosomes are recycled to the -1 index after they move off of the simulated mRNAs. This model does not consider binding energies of the ribosomes to sequences in the gene of interest.\n",
    "__References:__\n",
    "1. https://book.bionumbers.org/what-is-faster-transcription-or-translation\n",
    "2. https://www.columbia.edu/~ks20/stochastic-I/stochastic-I-MCI.pdf\n",
    "3. https://github.com/gvanderheide/discreteMarkovChain\n",
    "\n",
    "__Assumptions:__\n",
    "1. Initiation when ribosome reads first codon _(AUG in given gene file)_\n",
    "2. Elongation in-between\n",
    "3. Ribosome moves one triplet with each elongation step\n",
    "4. Ribosomes are recycled once they move off of the gene\n",
    "5. If any pair of ribosomes' positions are under 30 nucleotides apart, the ribosome closer to AUG is not allowed to move until its position $\\geq$ 30 nucleotides from the next further-along ribosome\n",
    "6. Termination when ribosome reads UAG, UAA, or UGA _(in next update)_\n",
    "7. Length of gene in graph and calculations is in codons, not nucleotides\n",
    "8. Position refers to the center position of the ribosome\n",
    "9. The mRNAs only consist of ORFs of the gene of interest _(may change in next update)_\n",
    "10. How quickly or if initiation increases, decreases _(sec.)_ _(constant rate)_\n",
    "11. How quickly or if elongation increases, decreases _(sec.)_ _(constant rate)_\n",
    "\n",
    "__Tunable Parameters:__\n",
    "1. Initiation rate\n",
    "2. Elongation rate\n",
    "3. Number of ribosomes on a single mRNA\n",
    "4. Number of mRNAs\n",
    "5. Probability of a ribosome moving from its current position\n",
    "6. Size of a ribosome (in nucleotides)\n",
    "7. How long the simulated run is for (sec.)\n",
    "\n",
    "(Ribosome reads 20$codon\\over{s.}$ at maximum elongation rate (1).)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "import numpy as np\n",
    "import math\n",
    "import sys\n",
    "import os\n",
    "from os import path\n",
    "import argparse\n",
    "from itertools import chain\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import *\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "plt.style.use('ggplot')\n",
    "#from discreteMarkovChain import markovChain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "global gene_length\n",
    "global mRNA\n",
    "global ribo_size\n",
    "global fname\n",
    "\n",
    "\"\"\"\n",
    "Read in the file containing the gene of interest and determine whether or not it is the correct file type \".txt\" and\n",
    "initialize the mRNA, occupancy, and ribosome_list lists.\n",
    "\"\"\"\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument('filename')\n",
    "args = parser.parse_args('sample.txt'.split())\n",
    "ribosome_list = [] # storage for all ribosomes in a single mRNA\n",
    "gene_length = 0\n",
    "mRNA = []\n",
    "occupancy = [] # storage for all ribosomes in all mRNAs\n",
    "fname = args.filename\n",
    "if not fname.lower().endswith(('.fasta','.txt')):\n",
    "    parser.error(\"file must be a .txt or FASTA file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates the ribosome object which stores the ribosome's position and how long\n",
    "    it has been at its current position. The counter is initially set to zero.\n",
    "\"\"\"\n",
    "class ribosome:\n",
    "    \"\"\"\n",
    "    Description: Creates a new ribosome object with initial position -1 and counter 0.\n",
    "    Inputs: None\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        self.position = -1\n",
    "        self.counter = 0\n",
    "    \"\"\"\n",
    "    Description: Changes the value of the ribosome's position.\n",
    "    Inputs: new_position - the position the ribosome has been moved to\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def set_position(self, new_position):\n",
    "        self.position = new_position\n",
    "    \"\"\"\n",
    "    Description: Changes the value of the ribosome object's counter variable given a new value.\n",
    "    Inputs: new_count - number of timesteps ribosome has been at its current position\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def set_counter(self, new_count):\n",
    "        self.counter = new_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Counts number of nucleotides in the gene of interest and creates an \"mRNA\" list of the same size\n",
    "    that will be used to note ribosome positions on.\n",
    "Inputs: gene_file - a .txt file that contains the mRNA base pairs of a given gene.\n",
    "Return: gene_length - the length of the gene passed in, in nucleotides\n",
    "        mRNA - a list corresponding with the length of the (gene_file in codons)\n",
    "\"\"\"\n",
    "def read_gene(gene_file):\n",
    "    global gene_length\n",
    "    global mRNA\n",
    "    temp_list = []\n",
    "    mRNA = []\n",
    "    \n",
    "    try:\n",
    "        with open(gene_file) as inputFileHandle:\n",
    "            line_list = [lines.split() for lines in inputFileHandle]  # extract lines\n",
    "            while line_list:\n",
    "                temp_list.extend(line_list.pop(0))\n",
    "                while temp_list:\n",
    "                    mRNA.extend(temp_list.pop(0))\n",
    "            gene_length = int(len(mRNA)/3)\n",
    "                    \n",
    "            return inputFileHandle.read(), gene_length, mRNA\n",
    "    except IOError:\n",
    "        sys.stderr.write(\"read_gene - Error: Could not open {}\\n\".format(gene_file))\n",
    "        sys.exit(-1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates a steady state probability matrix that determines the likelihood that a ribosome\n",
    "    will move to the next codon or remain in its current position.\n",
    "Inputs: None\n",
    "Return: None\n",
    "\"\"\"\n",
    "def make_matrix():\n",
    "    global probability\n",
    "    global gene_length\n",
    "    \n",
    "    probs = np.zeros[(gene_length,gene_length)]\n",
    "    for col in range(gene_length):\n",
    "        for row in range(gene_length):\n",
    "            if row == col:\n",
    "                probs[row][col] = probability\n",
    "                if row == gene_length - 1:\n",
    "                    probs[row][0] = abs(1-probability)\n",
    "            if row + 1 == col:\n",
    "                probs[row][col] = abs(1-probability)\n",
    "    mc = markovChain(probs)\n",
    "    mc.computePi('linear')                                      # computation method of steady-state probabilities\n",
    "    steady_state_probs = mc.pi                                  # steady-state probabilities of moving into different nucleotides\n",
    "    #print('equilibrium state: {}'.format(steady_state_probs))\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Given the probability that the ribosome will move, the ribosome is evaluated to move\n",
    "    or remain in its current position.\n",
    "Inputs: ribosome - ribosome object trying to move\n",
    "Return: ribosome - same ribosome object with updated position and counter\n",
    "\"\"\"\n",
    "def move(ribosome):\n",
    "    global gene_length\n",
    "    global probability\n",
    "    \n",
    "    position = ribosome.position\n",
    "    if ribosome.position < gene_length:\n",
    "        moves = np.random.random() <= probability\n",
    "        if moves:\n",
    "            ribosome.position = position + 1\n",
    "            ribosome.counter = 1\n",
    "        else:\n",
    "            ribosome.counter += 1\n",
    "    return ribosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Graphs the ribosome occupancy of codons as a histogram with x-axis being the position (in codons)\n",
    "    and the y-axis being occupancy as a fraction of the total number of ribosomes simulated across all mRNA\n",
    "Inputs: ribos - the list containing all ribosomes\n",
    "Return: None\n",
    "\"\"\"\n",
    "def create_histogram(all_cules):\n",
    "    global ribo_size\n",
    "    global fname\n",
    "    global gene_length\n",
    "    \n",
    "    pos_array = [ribo.position for sim in all_cules for ribo in sim]\n",
    "    ax = plt.gca()\n",
    "    ax.set_title('Ribosome Frequency ' + os.path.splitext(str(fname))[0])\n",
    "    ax.set_yscale('log')\n",
    "    hist = plt.hist(pos_array, bins=range(min(pos_array)*100,gene_length + 10, 10), density=False)\n",
    "    plt.xlabel('Distance from AUG (codon)')\n",
    "    plt.ylabel('Ribosome Frequency')\n",
    "    plt.show()\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Determines whether or not the ribosomes on the simulated mRNA will collide, which ribosome is allowed to\n",
    "    move, and updates the positions and number of timesteps present at a given location of all ribosomes corresponding\n",
    "    to a simulated mRNA.\n",
    "Inputs: ribosomes - the number of ribosomes to be run on a given mRNA\n",
    "        time - the length of the simulation (sec.)\n",
    "        ribo_size - the size of the ribosome object (nucleotides)\n",
    "        probability - the probability that a ribosome will move forward from its current position\n",
    "        kI - initiation rate\n",
    "        kE - elongation rate\n",
    "        n_mRNA - total number of mRNAs\n",
    "Return: ribosome_list - list containing all ribosome objects corresponding to a given mRNA\n",
    "\"\"\"\n",
    "def simulate(ribosomes,time,ribo_size,prob,kI,kE):\n",
    "    global gene_length\n",
    "    global probability\n",
    "    ################################# fix overlaps\n",
    "    ################################# when ribosome wants to move, also check if correct distance from \n",
    "    probability = prob\n",
    "    ribosome_list = [ribosome() for complex in range(ribosomes)]\n",
    "    flag, gene_length, mRNA = read_gene(args.filename)\n",
    "    discrete_time = [t for t in range(time*10)] # x10 is used because it splits the simulation into 0.1sec. res.\n",
    "    for t_step in discrete_time:\n",
    "        for complex in range(len(ribosome_list) - 1): # for each ribosome...\n",
    "            step = ribosome_list[complex].counter #-------------------------- wrong because may never reach?\n",
    "            if t_step % kI == 0: # if initiation timestep\n",
    "                if step % kE == 0: # if elongation timestep\n",
    "                    if ribosome_list[complex].position == 0: # if at beginning of gene\n",
    "                        if complex == 0: # if first ribosome in the simulation\n",
    "                            if ribosome_list[complex].counter == 0: # if just arrived at beginning of gene\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance from other ribosome\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                elif ribosome_list[len(ribosome_list)-1].position == -1:\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                                #continue # move on to next ribosome\n",
    "                            else: # if been at position for at least one time step\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance from other ribosome\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                elif ribosome_list[len(ribosome_list)-1].position == -1:\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                        else: # if not the first ribosome but at beginning of gene\n",
    "                            if ribosome_list[complex].counter == 0: # if just arrived at beginning of gene\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance from other ribosome\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                elif ribosome_list[complex-1].position == -1:\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                                #continue # move on to next ribosome\n",
    "                            else:\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance from other ribosome\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                                elif ribosome_list[complex-1].position == -1:\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                    elif ribosome_list[complex].position == -1: # if not on gene yet\n",
    "                        if complex == 0: # if first ribosome in the simulation\n",
    "                            if ribosome_list[len(ribosome_list)-1].position == -1: # if last ribosome also at -1\n",
    "                                moves = np.random.random() <= probability\n",
    "                                if moves:\n",
    "                                    ribosome_list[complex].counter = 1\n",
    "                                    ribosome_list[complex].position = 0\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                            else: # if last ribosome not at -1 but somewhere on the mRNA\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance from the last ribosome\n",
    "                                    moves = np.random.random() <= probability\n",
    "                                    if moves:\n",
    "                                        ribosome_list[complex].counter = 1\n",
    "                                        ribosome_list[complex].position = 0\n",
    "                                    else: ribosome_list[complex].counter += 1\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                        else: # if not the first ribosome\n",
    "                            if ribosome_list[complex-1].position == -1: # if ribosome \"ahead\" of it is also at -1\n",
    "                                moves = np.random.random() <= probability\n",
    "                                if moves:\n",
    "                                    ribosome_list[complex].counter = 1\n",
    "                                    ribosome_list[complex].position = 0\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                            else: # if ribosome \"ahead\" of it is not at -1 but somewhere else on the mRNA\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance fom other ribosome\n",
    "                                    moves = np.random.random() <= probability\n",
    "                                    if moves:\n",
    "                                        ribosome_list[complex].counter = 1\n",
    "                                        ribosome_list[complex].position = 0\n",
    "                                    else: ribosome_list[complex].counter += 1\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                    else: # if somewhere on gene not at beginning\n",
    "                        if complex == 0: # if first ribosome in the simulation\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                            ribo_size-1: # if correct distance apart\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                                if ribosome_list[complex].position > gene_length:\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            else: ribosome_list[complex].counter += 1\n",
    "                        else: # if not the first ribosome\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                            ribo_size-1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            elif ribosome_list[complex-1].position == -1:\n",
    "                                    ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                            else: ribosome_list[complex].counter += 1\n",
    "                else: # if not an elongation timestep and only an initiation timestep\n",
    "                    if complex == 0: # if first ribosome in the simulation\n",
    "                        if ribosome_list[complex].position == -1:\n",
    "                            if ribosome_list[len(ribosome_list)-1].position == -1:\n",
    "                                moves = np.random.random() <= probability\n",
    "                                if moves:\n",
    "                                    ribosome_list[complex].counter = 1\n",
    "                                    ribosome_list[complex].position = 0\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                            else: # if only first ribosome is at -1\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                                ribo_size-1: # if correct distance apart\n",
    "                                    moves = np.random.random() <= probability # attempt to move\n",
    "                                    if moves:\n",
    "                                        ribosome_list[complex].counter = 1\n",
    "                                        ribosome_list[complex].position = 0\n",
    "                                    else: ribosome_list[complex].counter += 1\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                        else: ribosome_list[complex].counter += 1\n",
    "                    else: # if complex > 0 and only an initiation timestep\n",
    "                        if ribosome_list[complex].position == -1: # if ribosome was not initiated or is recycled\n",
    "                            if ribosome_list[complex-1].position == -1: # if both ribosomes are at -1\n",
    "                                moves = np.random.random() <= probability # attempt to move\n",
    "                                if moves:\n",
    "                                    ribosome_list[complex].counter = 1\n",
    "                                    ribosome_list[complex].position = 0\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                            else: # if only current ribosome is at -1\n",
    "                                if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                                ribo_size-1: # if it is the correct distance away from its neighbor to the \"right\"\n",
    "                                    moves = np.random.random() <= probability\n",
    "                                    if moves:\n",
    "                                        ribosome_list[complex].counter = 1\n",
    "                                        ribosome_list[complex].position = 0\n",
    "                                    else: ribosome_list[complex].counter += 1\n",
    "                                else: ribosome_list[complex].counter += 1\n",
    "                        else: ribosome_list[complex].counter += 1\n",
    "            else: # if not initiation step\n",
    "                if ribosome_list[complex].position == 0: # if at the beginning of an mRNA\n",
    "                    if step % kE == 0: # if an elongation step\n",
    "                        if complex == 0: # if first ribosome in the simulation\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                            ribo_size-1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                                if ribosome_list[complex].position > gene_length:\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            elif ribosome_list[complex-1].position == -1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            else:\n",
    "                                ribosome_list[complex].counter += 1\n",
    "                        else: # if not the first ribosome in the simulation\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                            ribo_size-1: # if correct distance away from other ribosomes\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            elif ribosome_list[complex-1].position == -1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            else:\n",
    "                                ribosome_list[complex].counter += 1\n",
    "                    else:\n",
    "                        ribosome_list[complex].counter += 1 # if not an elongation step then increment amount of time spent in current state\n",
    "                elif ribosome_list[complex].position > 0: # if away from the beginning of the mRNA\n",
    "                    if step % kE == 0: # if elongation step\n",
    "                        if complex == 0:\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[len(ribosome_list)-1].position) > \\\n",
    "                            ribo_size-1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            elif ribosome_list[complex-1].position == -1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            else:\n",
    "                                ribosome_list[complex].counter += 1\n",
    "                        else: # if not the first ribosome\n",
    "                            if abs(ribosome_list[complex].position-ribosome_list[complex-1].position) > \\\n",
    "                            ribo_size-1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            elif ribosome_list[complex-1].position == -1:\n",
    "                                ribosome_list[complex] = move(ribosome_list[complex]) # attempt to move\n",
    "                                if ribosome_list[complex].position > gene_length: # if the ribosome leaves the ribosome it is recycled\n",
    "                                    ribosome_list[complex].position = -1\n",
    "                                    ribosome_list[complex].counter = 0\n",
    "                            else:\n",
    "                                ribosome_list[complex].counter += 1\n",
    "                    else:\n",
    "                        ribosome_list[complex].counter += 1 # if not an elongation step then increment amount of time spent in current state\n",
    "                else: # if at the -1 position and not initiation\n",
    "                    ribosome_list[complex].counter += 1 # if the ribosome was in state -1 then increment time spent in current state\n",
    "    \n",
    "    return ribosome_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates a list of lists containing\n",
    "Inputs: ribosomes - the number of ribosomes to be run on a given mRNA\n",
    "        time - the length of the simulation (sec.)\n",
    "        ribo_size - the size of the ribosome object (nucleotides)\n",
    "        probability - the probability that a ribosome will move forward from its current position\n",
    "        kI - initiation rate\n",
    "        kE - elongation rate\n",
    "        n_mRNA - total number of mRNAs\n",
    "Return: None\n",
    "\"\"\"\n",
    "def concatenate(ribosomes,time,ribo_size,prob,kI,kE,n_mRNA):    \n",
    "    occupancy = []\n",
    "    for mRNAs in range(n_mRNA):\n",
    "        occupancy.append(simulate(ribosomes,time,ribo_size,prob,kI,kE))\n",
    "    create_histogram(occupancy)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "848b15932f924a76b127f7c6d8e83bb5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>interactive</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "interactive(children=(IntSlider(value=25, description='n(ribosomes)', max=25, step=5), IntSlider(value=100, description='time (sec.)', max=10000, step=100), IntSlider(value=30, description='Ribo size (nt.)', max=35, min=25), FloatSlider(value=0.5, description='P(forward)', max=1.0, step=0.01), IntSlider(value=1, description='kI - try/sec.', max=10), IntSlider(value=1, description='kE - try/sec.', max=20), IntSlider(value=1, description='n(mRNA)', max=5000), Button(description='Run Interact', style=ButtonStyle()), Output()), _dom_classes=('widget-interact',))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initialize the sliders that control length of simulation, number of ribosomes, number of mRNA, size of mRNAs, kI, kE,\n",
    "    and the probability of moving to the next codon. Update the histogram when Run Interact button is pressed.\n",
    "\"\"\"\n",
    "\n",
    "initiation = widgets.IntSlider(min = 0, max = 10, value = 1, step = 1, description = 'kI - try/sec.');\n",
    "elongation = widgets.IntSlider(min = 0, max = 20, value = 1, step = 1, description = 'kE - try/sec.');\n",
    "prob_move = widgets.FloatSlider(min = 0.00, max = 1.00, value = 0.50, step = 0.01, description = 'P(forward)');\n",
    "sizes = widgets.IntSlider(min = 25, max = 35, value = 30, step = 1, description = 'Ribo size (nt.)');\n",
    "secs = widgets.IntSlider(min = 0, max = 10000, value = 100, step = 100, description = 'time (sec.)');\n",
    "ribo = widgets.IntSlider(min = 0, max = 25, value = 25, step = 5, description = 'n(ribosomes)');\n",
    "num_cules = widgets.IntSlider(min = 0, max = 5000, value = 1, step = 1, description = 'n(mRNA)');\n",
    "interact_manual(concatenate,ribosomes=ribo, time=secs, ribo_size=sizes, prob=prob_move, kI=initiation, kE=elongation, n_mRNA=num_cules);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
