{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tunable Ribosome Occupancy Model\n",
    "_Author: Raghav Chanchani_\n",
    "\n",
    "_For the Zid Lab at UC San Diego_\n",
    "\n",
    "_Date last modified: 11/11/2018_\n",
    "#### A discrete-time single probability chain to simulate ribosome distribution along a given gene using the assumptions listed below. There is no final absorption state. Ribosomes are recycled to the -1 index after they move off of the simulated mRNAs. This model does not consider binding energies of the ribosomes to sequences in the gene of interest. Time is simulated to 0.01sec. precision.\n",
    "__References:__\n",
    "1. https://book.bionumbers.org/what-is-faster-transcription-or-translation\n",
    "2. https://www.columbia.edu/~ks20/stochastic-I/stochastic-I-MCI.pdf\n",
    "3. https://github.com/gvanderheide/discreteMarkovChain\n",
    "\n",
    "__Assumptions:__\n",
    "1. Initiation when ribosome reads first codon _(AUG in given .txt or .fasta gene file)_\n",
    "2. Elongation in-between\n",
    "3. Ribosome moves one triplet with each elongation step\n",
    "4. Ribosomes are recycled once they move off of the gene\n",
    "5. If any pair of ribosomes' positions are under appropriate distance apart, then the ribosome closer to AUG is not allowed to move until its position $\\geq$ the distance threshold from the next ribosome\n",
    "6. Termination when ribosome reads UAG, UAA, or UGA _(in next update)_\n",
    "7. Length of gene in graph and calculations is in codons, not nucleotides\n",
    "8. Position refers to the center position of the ribosome\n",
    "9. Each second is divided into 100ths to allow for more accurate reflection of kI and kE values\n",
    "10. Calculations of $\\text{time res.}\\over{kI\\text{ or }kE}$ are \"floored\" (e.g. math.floor($100\\over{kI}$)) in determining when to attempt initiation and elongation\n",
    "\n",
    "__Tunable Parameters:__\n",
    "1. Initiation rate\n",
    "2. Elongation rate\n",
    "3. Number of ribosomes on a single mRNA\n",
    "4. Number of mRNAs\n",
    "5. Probability of a ribosome moving from its current position\n",
    "6. Size of a ribosome (in nucleotides)\n",
    "7. How long the simulated run is for (sec.)\n",
    "8. Time until initiation rate becomes 0 _(occurs stepwise)_\n",
    "9. Time until elongation rate becomes 0 _(occurs stepwise)_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "import numpy as np\n",
    "import math\n",
    "import sys\n",
    "import os\n",
    "from os import path\n",
    "import argparse\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import *\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "plt.style.use('ggplot')\n",
    "#from discreteMarkovChain import markovChain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Read in the file containing the gene of interest and determine whether or not it is the correct file type \".txt\" and\n",
    "initialize the mRNA, occupancy, and ribosome_list lists.\n",
    "\"\"\"\n",
    "global gene_length\n",
    "global mRNA\n",
    "global fname\n",
    "global res\n",
    "global codon_size\n",
    "\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument('filename')\n",
    "args = parser.parse_args('Pab1.txt'.split())\n",
    "ribosome_list = [] # storage for all ribosomes in a single mRNA\n",
    "gene_length = 0\n",
    "res = 100 # time resolution (the number of divisions made of a second)\n",
    "codon_size = 3\n",
    "mRNA = []\n",
    "occupancy = [] # storage for all ribosomes in all mRNAs\n",
    "fname = args.filename\n",
    "if not fname.lower().endswith(('.txt')):\n",
    "    parser.error(\"Input file must be a .txt file\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates the ribosome object which stores the ribosome's position and how long\n",
    "    it has been at its current position. The counter is initially set to zero.\n",
    "\"\"\"\n",
    "class ribosome:\n",
    "    \"\"\"\n",
    "    Description: Creates a new ribosome object with initial position -1 and counter 0.\n",
    "    Inputs: None\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        self.position = -1\n",
    "        self.counter = 0\n",
    "    \"\"\"\n",
    "    Description: Changes the value of the ribosome's position.\n",
    "    Inputs: new_position - the position the ribosome has been moved to\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def set_position(self, new_position):\n",
    "        self.position = new_position\n",
    "    \"\"\"\n",
    "    Description: Changes the value of the ribosome object's counter variable given a new value.\n",
    "    Inputs: new_count - number of timesteps ribosome has been at its current position\n",
    "    Return: None\n",
    "    \"\"\"\n",
    "    def set_counter(self, new_count):\n",
    "        self.counter = new_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Counts number of nucleotides in the gene of interest and creates an \"mRNA\" list of the same size\n",
    "    that will be used to note ribosome positions on.\n",
    "Inputs: gene_file - a .txt file that contains the mRNA base pairs of a given gene.\n",
    "Return: gene_length - the length of the gene passed in, in nucleotides\n",
    "        mRNA - a list corresponding with the length of the (gene_file in codons)\n",
    "\"\"\"\n",
    "def read_gene(gene_file):\n",
    "    global gene_length\n",
    "    global mRNA\n",
    "    global codon_size\n",
    "    global target\n",
    "    temp_list = []\n",
    "    mRNA = []\n",
    "\n",
    "    try:\n",
    "        with open(gene_file) as inputFileHandle:\n",
    "            line_list = [lines.split() for lines in inputFileHandle]  # extract lines\n",
    "            while line_list:\n",
    "                temp_list.extend(line_list.pop(0))\n",
    "                while temp_list:\n",
    "                    if ('>') in temp_list[0]:\n",
    "                        target = temp_list[0].strip('>')\n",
    "                    mRNA.extend(temp_list.pop(0))\n",
    "            gene_length = int(len(mRNA)/codon_size)\n",
    "            return inputFileHandle.read(), gene_length, mRNA\n",
    "    except IOError:\n",
    "        sys.stderr.write(\"read_gene - Error: Could not open {}\\n\".format(gene_file))\n",
    "        sys.exit(-1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates a steady state probability matrix that determines the likelihood that a ribosome\n",
    "    will move to the next codon or remain in its current position.\n",
    "Inputs: None\n",
    "Return: None\n",
    "\"\"\"\n",
    "def make_matrix():\n",
    "    global probability\n",
    "    global gene_length\n",
    "    \n",
    "    probs = np.zeros[(gene_length,gene_length)]\n",
    "    for col in range(gene_length):\n",
    "        for row in range(gene_length):\n",
    "            if row == col:\n",
    "                probs[row][col] = probability\n",
    "                if row == gene_length - 1:\n",
    "                    probs[row][0] = abs(1-probability)\n",
    "            if row + 1 == col:\n",
    "                probs[row][col] = abs(1-probability)\n",
    "    mc = markovChain(probs)\n",
    "    # computation method of steady-state probabilities\n",
    "    mc.computePi('linear')\n",
    "    # steady-state probabilities of moving into different nucleotides\n",
    "    steady_state_probs = mc.pi\n",
    "    print('equilibrium state: {}'.format(steady_state_probs))\n",
    "\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Given the probability that the ribosome will move, the ribosome is evaluated to move\n",
    "    or remain in its current position.\n",
    "Inputs: ribosome - ribosome object trying to move\n",
    "Return: ribosome - same ribosome object with updated position and counter\n",
    "\"\"\"\n",
    "def move(ribosome):\n",
    "    global gene_length\n",
    "    global probability\n",
    "    \n",
    "    position = ribosome.position\n",
    "    if ribosome.position <= gene_length:\n",
    "        moves = np.random.random() <= probability\n",
    "        if moves:\n",
    "            ribosome.position = position + 1\n",
    "            ribosome.counter = 1\n",
    "        else:\n",
    "            ribosome.counter += 1\n",
    "    return ribosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Graphs the ribosome occupancy of codons as a histogram with x-axis being the position (in codons)\n",
    "    and the y-axis being occupancy as a fraction of the total number of ribosomes simulated across all mRNA\n",
    "Inputs: ribos - the list containing all ribosomes\n",
    "Return: None\n",
    "\"\"\"\n",
    "def create_histogram(all_cules):\n",
    "    global fname\n",
    "    global gene_length\n",
    "    global res\n",
    "    global target\n",
    "    \n",
    "    pos_array = [ribo.position for sim in all_cules for ribo in sim]\n",
    "    ax = plt.gca()\n",
    "    ax.set_title('Ribosome Frequency ' + target)\n",
    "    #ax.set_yscale('log')\n",
    "    hist = plt.hist(pos_array, bins=range(0,gene_length + 1,5),density=False)\n",
    "    plt.xlabel('Distance from AUG (codons)')\n",
    "    plt.ylabel('Ribosome Frequency')\n",
    "    plt.show()\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def increase(start, end, curr, rise, rate, default):\n",
    "    run = abs(start - end)\n",
    "    if run == 0:\n",
    "        if curr == start:\n",
    "            rate = rise\n",
    "    else:\n",
    "        slp = rise / run\n",
    "        # increase from \"0\" to slider\n",
    "        if curr >= start and curr <= end:\n",
    "            if curr == end:\n",
    "                rate = rise\n",
    "            else:\n",
    "                rate = ((slp * curr) + (-slp * start))\n",
    "                if rate == 0:\n",
    "                    rate = default\n",
    "\n",
    "    return rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decrease(start, end, curr, fall, rate, default):\n",
    "    run = abs(start - end)\n",
    "    if run == 0:\n",
    "        if curr == start:\n",
    "            rate = default\n",
    "    else:\n",
    "        slp = fall / run\n",
    "        # decrease from slider to \"0\"\n",
    "        if curr >= start and curr <= end:\n",
    "            if curr == end:\n",
    "                rate = default\n",
    "            else:\n",
    "                rate = ((slp * curr) + (-slp * end))\n",
    "                if rate == 0:\n",
    "                    rate = default\n",
    "    \n",
    "    return rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Determines whether or not the ribosomes on the simulated mRNA will collide, which ribosome is allowed to\n",
    "    move, and updates the positions and number of timesteps present at a given location of all ribosomes corresponding\n",
    "    to a simulated mRNA.\n",
    "Inputs: ribosomes - the number of ribosomes to be run on a given mRNA\n",
    "        time - the length of the simulation (sec.)\n",
    "        ribo_size - the size of the ribosome object (nucleotides)\n",
    "        probability - the probability that a ribosome will move forward from its current position\n",
    "        kI - initiation rate\n",
    "        kE - elongation rate\n",
    "        n_mRNA - total number of mRNAs\n",
    "Return: ribosome_list - list containing all ribosome objects corresponding to a given mRNA\n",
    "\"\"\"\n",
    "def simulate(ribosomes, time, ribo_size, n_mRNA, prob, kI, kE, t_st_init, t_st_elo,\n",
    "               t_crit_st_init, t_crit_st_elo, t_stp_init, t_stp_elo, t_crit_stp_init,\n",
    "             t_crit_stp_elo):\n",
    "    global gene_length\n",
    "    global probability\n",
    "    global res\n",
    "    global codon_size\n",
    "    probability = prob\n",
    "    dead = res ** -codon_size\n",
    "    ribo_size = math.floor(ribo_size / codon_size)\n",
    "    if kE == 0:\n",
    "        slider_kE = dead\n",
    "    else:\n",
    "        slider_kE = kE\n",
    "    kE = dead\n",
    "    if kI == 0:\n",
    "        slider_kI = dead\n",
    "    else:\n",
    "        slider_kI = kI\n",
    "    kI = dead\n",
    "    # list of user-defined number of ribosomes\n",
    "    ribosome_list = [ribosome() for complex in range(ribosomes)]\n",
    "    flag, gene_length, mRNA = read_gene(args.filename)\n",
    "    # splits the simulation into 0.01sec. resolution\n",
    "    discrete_time = [t for t in range(time * res)]\n",
    "    for t_step in discrete_time:\n",
    "        # Increase initiation rate starting at t_st_init until t_crit_init to kI\n",
    "        init_st_run = t_crit_st_init - t_st_init\n",
    "        if init_st_run == 0:\n",
    "            if t_step == t_st_init:\n",
    "                kI = slider_kI\n",
    "        else:\n",
    "            init_rise = slider_kI\n",
    "            # calculate slope to adjust kI by for each t_step in range\n",
    "            init_up_slp = init_rise / init_st_run\n",
    "            # increase initiation from \"0\" to slider kI value\n",
    "            if t_step >= t_st_init and t_step <= t_crit_st_init:\n",
    "                if t_step == t_crit_st_init:\n",
    "                    kI = slider_kI\n",
    "                else:\n",
    "                    kI = ((init_up_slp * (t_step)) + (-init_up_slp * t_st_init))\n",
    "                    if kI == 0:\n",
    "                        kI = dead\n",
    "\n",
    "        # Decrease initiation rate starting at t_stp_init until t_crit_stp_init to 0\n",
    "        init_fall = -slider_kI\n",
    "        init_stp_run = t_crit_stp_init - t_stp_init\n",
    "        if init_stp_run == 0:\n",
    "            if t_step == t_stp_init:\n",
    "                kI = dead\n",
    "            else:\n",
    "                # decrease initiation from slider kI to effectively 0\n",
    "                if t_step <= t_stp_init and t_step >= t_crit_stp_init:\n",
    "                    if t_step == t_crit_stp_init:\n",
    "                        kI = dead\n",
    "        else:\n",
    "            init_dwn_slp = init_fall / init_stp_run\n",
    "            # decrease initiation from slider kI to effectively 0\n",
    "            if t_step <= t_stp_init and t_step >= t_crit_stp_init:\n",
    "                if t_step == t_crit_stp_init:\n",
    "                    kI = dead\n",
    "                else:\n",
    "                    kI = ((init_dwn_slp * (t_step)) + (-init_dwn_slp * t_crit_stp_init))\n",
    "                    if kI == 0:\n",
    "                        kI = dead\n",
    "        \n",
    "        # Increase elongation rate starting at t_st_elo until t_crit_elo to kE\n",
    "        elo_st_run = t_crit_st_elo - t_st_elo\n",
    "        elo_rise = slider_kE\n",
    "        if elo_st_run == 0:\n",
    "            if t_step == t_st_elo:\n",
    "                kE = slider_kE\n",
    "        else:\n",
    "            elo_up_slp = elo_rise / elo_st_run\n",
    "            # increase elongation from \"0\" to slider kE value\n",
    "            if t_step >= t_st_elo and t_step <= t_crit_st_elo:\n",
    "                if t_step == t_crit_st_elo:\n",
    "                    kE = slider_kE\n",
    "                else:\n",
    "                    kE = ((elo_up_slp * (t_step)) + (-elo_up_slp * t_st_elo))\n",
    "                    if kE == 0:\n",
    "                        kE = dead\n",
    "        \n",
    "        # Decrease elongation rate starting at t_stp_init until t_crit_stp_init to 0\n",
    "        elo_fall = -slider_kE\n",
    "        elo_stp_run = t_crit_stp_elo - t_stp_elo\n",
    "        if elo_stp_run == 0:\n",
    "            if t_step == t_stp_elo:\n",
    "                kE = dead\n",
    "            else:\n",
    "                # decrease elongation from slider kE to \"0\"\n",
    "                if t_step <= t_stp_elo and t_step >= t_crit_stp_elo:\n",
    "                    if t_step == t_crit_stp_elo:\n",
    "                        kE = dead\n",
    "        else:\n",
    "            elo_dwn_slp = elo_fall / elo_stp_run\n",
    "            # decrease elongation from slider kE to \"0\"\n",
    "            if t_step <= t_stp_elo and t_step >= t_crit_stp_elo:\n",
    "                if t_step == t_crit_stp_elo:\n",
    "                    kE = dead\n",
    "                else:\n",
    "                    kE = ((elo_dwn_slp * (t_step)) + (-elo_dwn_slp * t_crit_stp_elo))\n",
    "                    if kE == 0:\n",
    "                        kE = dead\n",
    "                        \n",
    "        # each ribosome...\n",
    "        for complex in range(len(ribosome_list)):\n",
    "            # compared ribosome counter variable\n",
    "            n_ribo = 0\n",
    "            step = ribosome_list[complex].counter\n",
    "            if t_step % int(res / kI) == 0:\n",
    "                if step % int(res / kE) == 0:\n",
    "                    for ribo in ribosome_list:\n",
    "                        if ribosome_list.index(ribo) == complex:\n",
    "                            n_ribo += 1\n",
    "                        elif ribo.position > ribosome_list[complex].position:\n",
    "                            if ribo.position - ribosome_list[complex].position > ribo_size: # can move\n",
    "                                n_ribo += 1\n",
    "                            else: \n",
    "                                break\n",
    "                        elif ribo.position < ribosome_list[complex].position:\n",
    "                            n_ribo += 1\n",
    "                        elif ribo.position == -1:\n",
    "                            n_ribo += 1\n",
    "                        else:\n",
    "                            n_ribo += 1\n",
    "                    if n_ribo == len(ribosome_list):\n",
    "                        ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                        if ribosome_list[complex].position > gene_length:\n",
    "                            ribosome_list[complex].position = -1\n",
    "                            ribosome_list[complex].counter = 1\n",
    "                    else: ribosome_list[complex].counter += 1\n",
    "                # if only initiation timestep\n",
    "                else:\n",
    "                    if ribosome_list[complex].position == -1:\n",
    "                        for ribo in ribosome_list:\n",
    "                            if ribosome_list.index(ribo) == complex:\n",
    "                                n_ribo += 1\n",
    "                            elif ribo.position > ribosome_list[complex].position:\n",
    "                                if ribo.position - ribosome_list[complex].position > ribo_size: # can move\n",
    "                                    n_ribo += 1\n",
    "                                else: \n",
    "                                    break\n",
    "                            elif ribo.position < ribosome_list[complex].position:\n",
    "                                n_ribo += 1\n",
    "                            elif ribo.position == -1:\n",
    "                                n_ribo += 1\n",
    "                            else:\n",
    "                                n_ribo += 1\n",
    "                        if n_ribo == len(ribosome_list):\n",
    "                            ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                            if ribosome_list[complex].position > gene_length:\n",
    "                                ribosome_list[complex].position = -1\n",
    "                                ribosome_list[complex].counter = 1\n",
    "                        else: ribosome_list[complex].counter += 1\n",
    "                    else: ribosome_list[complex].counter += 1\n",
    "            # if not initiation timestep\n",
    "            else:\n",
    "                # if elongation timestep but not initiation timestep\n",
    "                if step % int(res / kE) == 0:\n",
    "                    if ribosome_list[complex].position == -1:\n",
    "                        ribosome_list[complex].counter += 1\n",
    "                    else:\n",
    "                        for ribo in ribosome_list:\n",
    "                            if ribosome_list.index(ribo) == complex:\n",
    "                                n_ribo += 1\n",
    "                            elif ribo.position > ribosome_list[complex].position:\n",
    "                                if ribo.position - ribosome_list[complex].position > ribo_size: # can move\n",
    "                                    n_ribo += 1\n",
    "                                else: \n",
    "                                    break\n",
    "                            elif ribo.position < ribosome_list[complex].position:\n",
    "                                n_ribo += 1\n",
    "                            elif ribo.position == -1:\n",
    "                                n_ribo += 1\n",
    "                            else:\n",
    "                                n_ribo += 1\n",
    "                        if n_ribo == len(ribosome_list):\n",
    "                            ribosome_list[complex] = move(ribosome_list[complex])\n",
    "                            if ribosome_list[complex].position > gene_length:\n",
    "                                ribosome_list[complex].position = -1\n",
    "                                ribosome_list[complex].counter = 1\n",
    "                        else: ribosome_list[complex].counter += 1\n",
    "                # if not initiation nor elongation timestep\n",
    "                else:\n",
    "                    ribosome_list[complex].counter += 1\n",
    "    return ribosome_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Description: Creates a list of lists containing\n",
    "Inputs: ribosomes - the number of ribosomes to be run on a given mRNA\n",
    "        time - the length of the simulation (sec.)\n",
    "        ribo_size - the size of the ribosome object (nucleotides)\n",
    "        probability - the probability that a ribosome will move forward from its current position\n",
    "        kI - initiation rate\n",
    "        kE - elongation rate\n",
    "        n_mRNA - total number of mRNAs\n",
    "Return: None\n",
    "\"\"\"\n",
    "def concatenate(ribosomes, time, ribo_size, n_mRNA, prob, kI, kE, t_st_init, t_st_elo,\n",
    "               t_crit_st_init, t_crit_st_elo, t_stp_init, t_stp_elo, t_crit_stp_init, t_crit_stp_elo):\n",
    "    global gene_length\n",
    "    occupancy = []\n",
    "    \n",
    "    for mRNAs in range(n_mRNA):\n",
    "        #occupancy.append(simulate(ribosomes,time,ribo_size,prob,kI,kE,t_init,t_elo,t_init_st,t_elo_st))\n",
    "        occupancy.append(simulate(ribosomes, time,\n",
    "                                  ribo_size, n_mRNA,\n",
    "                                  prob, kI,\n",
    "                                  kE, t_st_init,\n",
    "                                  t_st_elo, t_crit_st_init,\n",
    "                                  t_crit_st_elo, t_stp_init,\n",
    "                                  t_stp_elo, t_crit_stp_init,\n",
    "                                  t_crit_stp_elo))\n",
    "    create_histogram(occupancy)\n",
    "    print(\"gene length: {} codons\".format(gene_length))\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "33edefb8cefe4fb480763793bf0b62af",
       "version_major": 2,
       "version_minor": 0
      },
      "text/html": [
       "<p>Failed to display Jupyter Widget of type <code>interactive</code>.</p>\n",
       "<p>\n",
       "  If you're reading this message in the Jupyter Notebook or JupyterLab Notebook, it may mean\n",
       "  that the widgets JavaScript is still loading. If this message persists, it\n",
       "  likely means that the widgets JavaScript library is either not installed or\n",
       "  not enabled. See the <a href=\"https://ipywidgets.readthedocs.io/en/stable/user_install.html\">Jupyter\n",
       "  Widgets Documentation</a> for setup instructions.\n",
       "</p>\n",
       "<p>\n",
       "  If you're reading this message in another frontend (for example, a static\n",
       "  rendering on GitHub or <a href=\"https://nbviewer.jupyter.org/\">NBViewer</a>),\n",
       "  it may mean that your frontend doesn't currently support widgets.\n",
       "</p>\n"
      ],
      "text/plain": [
       "interactive(children=(IntSlider(value=25, description='n(ribosomes)', max=25), IntSlider(value=100, description='time (sec.)', max=90000, step=10), IntSlider(value=30, description='Ribo size (nt.)', max=35, min=25), IntSlider(value=10, description='n(mRNA)', max=5000), FloatSlider(value=0.5, description='P(forward)', max=1.0, step=0.01), IntSlider(value=1, description='kI - try/sec.', max=10), IntSlider(value=20, description='kE - try/sec.', max=20), IntSlider(value=50, description='sec. delay kI', max=90001), IntSlider(value=50, description='sec. delay kE', max=90001), IntSlider(value=50, description='sec. until kI', max=90001), IntSlider(value=50, description='sec. until kE', max=90001), IntSlider(value=20, description='kI start decay', max=90001), IntSlider(value=20, description='kE start decay', max=90001), IntSlider(value=50, description='sec. to kI = 0', max=90001), IntSlider(value=50, description='sec. to kE = 0', max=90001), Button(description='Run Interact', style=ButtonStyle()), Output()), _dom_classes=('widget-interact',))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Initialize the sliders that control length of simulation, number of ribosomes, number of mRNA, size of mRNAs, kI, kE,\n",
    "    and the probability of moving to the next codon. Update the histogram when Run Interact button is pressed.\n",
    "\"\"\"\n",
    "initiation = widgets.IntSlider(min = 0, max = 10, value = 1, step = 1, description = 'kI - try/sec.');\n",
    "init_stop_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. to kI = 0');\n",
    "elongation = widgets.IntSlider(min = 0, max = 20, value = 20, step = 1, description = 'kE - try/sec.');\n",
    "elo_stop_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. to kE = 0');\n",
    "init_start_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. delay kI');\n",
    "elo_start_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. delay kE');\n",
    "init_crit_inc_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. until kI');\n",
    "elo_crit_inc_time = widgets.IntSlider(min = 0, max = 90001, value = 50, step = 1, description = 'sec. until kE');\n",
    "prob_move = widgets.FloatSlider(min = 0.00, max = 1.00, value = 0.50, step = 0.01, description = 'P(forward)');\n",
    "sizes = widgets.IntSlider(min = 25, max = 35, value = 30, step = 1, description = 'Ribo size (nt.)');\n",
    "secs = widgets.IntSlider(min = 0, max = 90000, value = 100, step = 10, description = 'time (sec.)');\n",
    "ribo = widgets.IntSlider(min = 0, max = 25, value = 25, step = 1, description = 'n(ribosomes)');\n",
    "num_cules = widgets.IntSlider(min = 0, max = 5000, value = 10, step = 1, description = 'n(mRNA)');\n",
    "init_decay_time = widgets.IntSlider(min = 0, max = 90001, value = 20, step = 1, description = 'kI start decay');\n",
    "elo_decay_time = widgets.IntSlider(min = 0, max = 90001, value = 20, step = 1, description = 'kE start decay');\n",
    "interact_manual(concatenate,\n",
    "                ribosomes = ribo,\n",
    "                time = secs,\n",
    "                ribo_size = sizes,\n",
    "                n_mRNA = num_cules,\n",
    "                prob = prob_move,\n",
    "                kI = initiation,\n",
    "                kE = elongation,\n",
    "                t_st_init = init_start_time,\n",
    "                t_st_elo = elo_start_time,\n",
    "                t_crit_st_init = init_crit_inc_time,\n",
    "                t_crit_st_elo = elo_crit_inc_time,\n",
    "                t_stp_init = init_decay_time,\n",
    "                t_stp_elo = elo_decay_time,\n",
    "                t_crit_stp_init = init_stop_time,\n",
    "                t_crit_stp_elo = elo_stop_time);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
