{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21475045\n",
      "28030106\n",
      "593662\n"
     ]
    }
   ],
   "source": [
    "#Trims Fastq files of polyA tails for Bowtie Output\n",
    "\n",
    "file1 = open('D:/ZidLab/SequenceReads/Kyle_7_18_17/WT_5D_FP_S3_L003_R1_001.fastq' , 'r')\n",
    "output = open('D:/Bowtie/bowtie-1.1.2/WT_5DFP_7_17.fastq' , 'w')\n",
    "count = 0\n",
    "count1 = 0\n",
    "count2 = 0\n",
    "num = 0\n",
    "read = 0\n",
    "for line in file1 :\n",
    "        num+=1\n",
    "        if num ==2:  #The file has sequences on the 2nd line, every 4 lines\n",
    "                temp2 = line.rstrip()\n",
    "                temp26 = temp2[0:25] #Take the first 26 bps\n",
    "                tempminus1 = temp26.rstrip('A') #Strip all the A's\n",
    "                tempminus2 = tempminus1.rstrip('N') #Strip the next base (in case an A was misread\n",
    "                tempminusA = tempminus2.rstrip('A') #Strip more A's\n",
    "                Length = len(tempminusA)\n",
    "                tempTest = tempminusA\n",
    "                tempTest1 = tempTest[0:-1]\n",
    "                tempTest2 = tempTest1.rstrip('A')\n",
    "                LenChange = Length - len(tempTest2) #I'm double checking for extra runs of A's, not sure if its needed.\n",
    "                if LenChange > 6 :\n",
    "                        count2 = count2 + 1\n",
    "                        read = 7\n",
    "                else:\n",
    "                        if Length > 17 :\n",
    "                                if Length > 23:\n",
    "                                        Sequence = tempminusA[0:23] #Only use the first 23bps for alignment\n",
    "                                        Length = len(Sequence)      #More likely to have misreads later in the sequence\n",
    "                                else:\n",
    "                                        Sequence = tempminusA\n",
    "                                acount = Sequence.count('A')\n",
    "                                temp4 = Length - 5\n",
    "                                if acount < temp4 :\n",
    "                                        count = count + 1\n",
    "                                        read = 0\n",
    "                                else:\n",
    "                                        read = 7\n",
    "                        else:\n",
    "                                read = 7\n",
    "                \n",
    "        \n",
    "        elif num == 4: #This is to get the Quality of the sequence read\n",
    "                count1+=1\n",
    "                num = 0\n",
    "                QualityT = line.rstrip()\n",
    "                Quality = QualityT[0:Length]\n",
    "                if read == 7:\n",
    "                        read = 0\n",
    "                else:\n",
    "                        output.write('@r'+str(count)+'/1.'+str(Length)+'\\n'+Sequence+'\\n'+'+\\n'+Quality+'\\n')\n",
    "                \n",
    "output.close()\n",
    "print count\n",
    "print count1\n",
    "print count2\n",
    "#Takes txt Solexa Sequence results and outputs them in fastq for\n",
    "# Took the first 26bp, removed the 3'AAAAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1617972\n"
     ]
    }
   ],
   "source": [
    "#Converts Bowtie Output to counts per chromosome site\n",
    "numerals = [\n",
    "        {'letter': 'M', 'value': 1000},\n",
    "        {'letter': 'D', 'value': 500},\n",
    "        {'letter': 'C', 'value': 100},\n",
    "        {'letter': 'L', 'value': 50},\n",
    "        {'letter': 'X', 'value': 10},\n",
    "        {'letter': 'V', 'value': 5},\n",
    "        {'letter': 'I', 'value': 1},\n",
    "    ]\n",
    "def roman_to_arabic(number):\n",
    "    index_by_letter = {}\n",
    "    for index in xrange(len(numerals)):\n",
    "        index_by_letter[numerals[index]['letter']] = index\n",
    "\n",
    "    result = 0\n",
    "    previous_value = None\n",
    "    for letter in reversed(number):\n",
    "        index = index_by_letter[letter]\n",
    "        value = numerals[index]['value']\n",
    "        if (previous_value is None) or (previous_value <= value):\n",
    "            result += value\n",
    "        else:\n",
    "            result -= value\n",
    "        previous_value = value\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "file1 = open('C:/Zid/SolexaDataAnalysis/Kyle/WT1DFP_7_17.map','r') \n",
    "file2 = open('C:/Zid/SolexaDataAnalysis/Kyle/output_WT1DFP.fq','r') \n",
    "output = open('C:/Zid/SolexaDataAnalysis/Kyle/wt1DFPAlignSortCount.txt','w')\n",
    "count2 = 0\n",
    "count3 =0\n",
    "temp = file1.readline()\n",
    "line1 = temp.rstrip() \n",
    "temp1 = line1.split('\\t')\n",
    "ChrTemp2 = temp1[2]\n",
    "if ChrTemp2 == \"MT\":\n",
    "    Chr = 17\n",
    "else:\n",
    "    Chr = roman_to_arabic(ChrTemp2)\n",
    "\n",
    "Name = temp1[0]\n",
    "NameLen = Name[-2:]\n",
    "Orien = temp1[1]\n",
    "if Orien == '-':\n",
    "    Posit = (int(temp1[3]))+(int(NameLen))  #didn't subtract 1 adjusting because it is one off\n",
    "else :\n",
    "    Posit = (int(temp1[3]))+1 #adjusting because it seems to be one off on the +Strand\n",
    "    \n",
    "\n",
    "dct_lst = [{\"Chr\":Chr,\"Posit\":Posit,\"Orien\": Orien}]#\"Seq\": 'a'}]\n",
    "\n",
    "for line in file1 :\n",
    "    count3 +=1\n",
    "    line1 = line.rstrip() \n",
    "    temp1 = line1.split('\\t')\n",
    "    ChrTemp2 = temp1[2]\n",
    "    if ChrTemp2 == \"MT\":\n",
    "        Chr = 17\n",
    "    else:\n",
    "        Chr = roman_to_arabic(ChrTemp2)\n",
    "    Name = temp1[0]\n",
    "    NameLen = Name[-2:]\n",
    "    Orien = temp1[1]\n",
    "    if Orien == '-':\n",
    "        Posit = (int(temp1[3]))+(int(NameLen))\n",
    "    elif Orien =='+' :\n",
    "        Posit = (int(temp1[3]))+1\n",
    "    temp_lst = [{\"Chr\":Chr,\"Posit\":Posit,\"Orien\": Orien}]#\"Seq\": 'a'}]#temp1[4]\n",
    "    dct_lst.extend(temp_lst)\n",
    "\n",
    "for line in file2 :    \n",
    "    count3 +=1\n",
    "    line1 = line.rstrip() \n",
    "    temp1 = line1.split('\\t')\n",
    "    ChrTemp2 = temp1[2]\n",
    "    if ChrTemp2 == \"MT\":\n",
    "        Chr = 17\n",
    "    else:\n",
    "        Chr = roman_to_arabic(ChrTemp2)\n",
    "    Name = temp1[0]\n",
    "    NameLen = Name[-2:]\n",
    "    Orien = temp1[1]\n",
    "    if Orien == '-':\n",
    "        Posit = (int(temp1[3]))+(int(NameLen))\n",
    "    elif Orien =='+' :\n",
    "        Posit = (int(temp1[3]))+1\n",
    "    temp_lst = [{\"Chr\":Chr,\"Posit\":Posit,\"Orien\": Orien}]#\"Seq\": 'a'}]#temp1[4]\n",
    "    dct_lst.extend(temp_lst)\n",
    "    \n",
    "from operator import itemgetter\n",
    "dct_lst.sort(key=itemgetter(\"Posit\"))\n",
    "\n",
    "from operator import itemgetter\n",
    "dct_lst.sort(key=itemgetter(\"Chr\"))\n",
    "\n",
    "\n",
    "count = 1\n",
    "Length = len(dct_lst)\n",
    "for i in range(len(dct_lst)) :\n",
    "    Chr = (dct_lst[i][\"Chr\"])\n",
    "    Posit = (dct_lst[i][\"Posit\"])\n",
    "    Orien = (dct_lst[i][\"Orien\"])\n",
    "    #Seq = (dct_lst[i][\"Seq\"])\n",
    "    if i == (Length-1):\n",
    "        output.write(Orien+'\\t'+str(Chr)+'\\t'+str(count)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n')\n",
    "    else:\n",
    "        OrienT2 = (dct_lst[i+1][\"Orien\"])\n",
    "        PositT2 = (dct_lst[i+1][\"Posit\"])\n",
    "        if Orien == OrienT2 :\n",
    "        \n",
    "            if Posit == PositT2 :\n",
    "                count = count + 1\n",
    "            else :\n",
    "                count2 = count + count2\n",
    "                output.write(Orien+'\\t'+str(Chr)+'\\t'+str(count)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n') #sense count\n",
    "                count = 1\n",
    "        else :\n",
    "            count2 = count + count2\n",
    "            output.write(Orien+'\\t'+str(Chr)+'\\t'+str(count)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n') #sense count\n",
    "            count = 1\n",
    "         \n",
    "output.close()  \n",
    "print count2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Converts accumulated sorted data to Mochiview format\n",
    "file1 = open('D:/ZidData/OSheaLab/Solexa/GluTime2/Aligning/AlignedFiles/5P_S2_9p35Align.txt','r')\n",
    "output = open('D:/ZidData/OSheaLab/MochiView/5Pp35_comb_Mochi.txt', 'w')\n",
    "output.write('SEQ_NAME\\tSTART\\tEND\\tSTRAND\\tCOUNT\\n')\n",
    "#temp7 = file1.readline()\n",
    "norm = 2.65 # Normalized to total alignable reads/3 million\n",
    "\n",
    "for line in file1:\n",
    "    temp = line.rstrip()\n",
    "    temp1 = temp.split('\\t') \n",
    "    name = \"Chr\"+temp1[1]\n",
    "       \n",
    "    PositTemp = int(temp1[3])\n",
    "    Strand = temp1[0]\n",
    "    if Strand == '+' :\n",
    "        count = (float(temp1[2]))/norm\n",
    "        Start = str(PositTemp)\n",
    "        \n",
    "    else :\n",
    "        tempc = (float(temp1[2]))/norm\n",
    "        count = tempc*(-1)\n",
    "        Start = str(PositTemp)\n",
    "        \n",
    "    output.write(name+'\\t'+Start+'\\t'+Start+'\\t'+Strand+'\\t'+str(count)+ '\\n')  #antisense count\n",
    "\n",
    "\n",
    "output.close()\n",
    "file1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "960448.0\n"
     ]
    }
   ],
   "source": [
    "#Converts accumulated sorted data at the genome level to gene features\n",
    "\n",
    "file1 = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/3Pp35CombinedAlignSortCount.txt','r')\n",
    "output = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/3Pp35Feat.txt', 'w')\n",
    "AlignCountTotal = 0.0\n",
    "\n",
    "for line in file1:\n",
    "    temp = line.rstrip()\n",
    "    temp1 = temp.split('\\t') \n",
    "    Chr = int(temp1[1])\n",
    "    Posit = int(temp1[3])\n",
    "    Posit = Posit + 1\n",
    "    Orien = temp1[0]\n",
    "    count = temp1[2]\n",
    "    AlignCount = 0\n",
    "    \n",
    "    #Seq = temp1[5]\n",
    "    if Chr == 1 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND1.txt','r')\n",
    "    elif Chr == 2 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND2.txt','r')\n",
    "    elif Chr == 3 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND3.txt','r')\n",
    "    elif Chr == 4 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND4.txt','r')\n",
    "    elif Chr == 5 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND5.txt','r')\n",
    "    elif Chr == 6 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND6.txt','r')\n",
    "    elif Chr == 7 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND7.txt','r')\n",
    "    elif Chr == 8 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND8.txt','r')\n",
    "    elif Chr == 9 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND9.txt','r')\n",
    "    elif Chr == 10 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND10.txt','r')\n",
    "    elif Chr == 11 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND11.txt','r')\n",
    "    elif Chr == 12 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND12.txt','r')\n",
    "    elif Chr == 13 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND13.txt','r')\n",
    "    elif Chr == 14 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND14.txt','r')\n",
    "    elif Chr == 15 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND15.txt','r')\n",
    "    elif Chr == 16 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND16.txt','r')\n",
    "    elif Chr == 17 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND17.txt','r')    \n",
    "    \n",
    "    for line2 in file2 :\n",
    "        temp3 = line2.rstrip()\n",
    "        temp2 = temp3.split('\\t')\n",
    "        Chr2 = int(temp2[1])\n",
    "        if Chr == Chr2: #checking if the same chromosome\n",
    "            if Orien == '-' :  # C orientation\n",
    "                name = temp2[0]\n",
    "                Orien2 = temp2[4]\n",
    "                if Orien2 == '-' :  #also C orientation\n",
    "                    tStrt2 = int(temp2[3])\n",
    "                    tEnd2 = int(temp2[2])\n",
    "                    Strt2 = tStrt2+16\n",
    "                    End2 = tEnd2+13\n",
    "                                    \n",
    "                    if End2 < Posit < Strt2 :\n",
    "                        AlignCount = AlignCount + 1\n",
    "                        output.write(name+'\\t'+str(tStrt2)+'\\t'+str(tEnd2)+'\\t'+str(Posit)+'\\t'+count+'\\n') #sense count\n",
    "                \n",
    "                  \n",
    "            else :          # W orientation\n",
    "                name = temp2[0]\n",
    "                Orien2 = temp2[4]\n",
    "                if Orien2 == '+' :  #also W orientation\n",
    "                    tStrt2 = int(temp2[2])\n",
    "                    tEnd2 = int(temp2[3])\n",
    "                    Strt2 = tStrt2-16\n",
    "                    End2 = tEnd2-13\n",
    "\n",
    "                    if Strt2 < Posit < End2 :\n",
    "                        AlignCount = AlignCount + 1\n",
    "                        output.write(name+'\\t'+str(tStrt2)+'\\t'+str(tEnd2)+'\\t'+str(Posit)+ '\\t'+count+'\\n') #sense count\n",
    "                        AlignCountTotal = AlignCountTotal + float(count)\n",
    "    file2.close()\n",
    "print AlignCountTotal\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Outputs data to excel file with counts per gene\n",
    "import sys\n",
    "\n",
    "file1 = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/1Pp35Feat.txt','r')\n",
    "output = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/1Pp35Sorted.txt','w')\n",
    "temp = file1.readline()\n",
    "line1 = temp.rstrip() \n",
    "temp1 = line1.split('\\t')\n",
    "Name = temp1[0]\n",
    "ChrT = Name[2]\n",
    "Chr = '1'\n",
    "dct_lst = [{\"name\":temp1[0],\"Start\":temp1[1],\"End\":\"7236\",\"Posit\":temp1[3],\"Count\":temp1[4]}]\n",
    "\n",
    "for line in file1 :\n",
    "   temp = line.rstrip()\n",
    "   temp1 = temp.split('\\t')\n",
    "     \n",
    "   temp_lst = [{\"name\":temp1[0],\"Start\":temp1[1],\"End\":temp1[2],\"Posit\":temp1[3],\"Count\":temp1[4]}]\n",
    "   dct_lst.extend(temp_lst)\n",
    "file1.close()\n",
    "\n",
    "#for line in file2:\n",
    "  \n",
    "#   temp = line.rstrip()\n",
    "#   temp1 = temp.split('\\t')  \n",
    "#   temp_lst = [{\"name\":temp1[0],\"Start\":temp1[1],\"End\":temp1[2],\"Posit\":temp1[3],\"Count\":temp1[4],\"ASCount\": temp1[5]}]\n",
    "#   dct_lst.extend(temp_lst)\n",
    "\n",
    "#file2.close()\n",
    "\n",
    "\n",
    "from operator import itemgetter\n",
    "dct_lst.sort(key=itemgetter(\"name\"))\n",
    "\n",
    "\n",
    "for i in range(len(dct_lst)) :\n",
    "   output.write(dct_lst[i][\"name\"]+'\\t'+dct_lst[i][\"Start\"]+'\\t'+dct_lst[i][\"End\"]+'\\t'+dct_lst[i][\"Posit\"]+'\\t'+dct_lst[i][\"Count\"]+'\\n') #sense count\n",
    "      \n",
    "output.close()\n",
    "\n",
    "file1 = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/1Pp35Sorted.txt')\n",
    "output = open('C:/Zid/SolexaDataAnalysis/EYp35Timecourse/1Pp35Counted.txt','w')\n",
    "output.write('Name\\tStrt\\tEnd\\tLength\\tCount\\tStallCount\\tCodeCount\\t200Count\\t200EndCount\\n')\n",
    "\n",
    "\n",
    "tName = 'YAL001C'\n",
    "tStrt = 147596\n",
    "tEnd = 151168\n",
    "tLength = 3573\n",
    "tCount = 0\n",
    "Count =0\n",
    "SCount = 0\n",
    "CCount = 0\n",
    "CountS = 0\n",
    "CountE = 0\n",
    "\n",
    "for line in file1:\n",
    "    temp = line.rstrip()\n",
    "    temp1 = temp.split('\\t')\n",
    "    Name = temp1[0]\n",
    "   \n",
    "    if Name[0] == 'Q':\n",
    "        Count = 0\n",
    "    elif Name[0] == 'Y':\n",
    "        Strt = int(temp1[1])\n",
    "        End = int(temp1[2])\n",
    "        Orien = Name[6]\n",
    "        Length = abs(Strt-End)\n",
    "        Posit = int(temp1[3])\n",
    "        tCount = int(temp1[4])\n",
    "        if tName == Name:\n",
    "            Count = Count + tCount\n",
    "            if Orien == 'W':\n",
    "                if Strt+20 > Posit and Strt-16 < Posit:\n",
    "                    SCount = SCount + tCount\n",
    "                elif Strt+220 > Posit and Strt-16 < Posit:\n",
    "                    CountS = CountS + tCount\n",
    "                elif End - 200 < Posit:\n",
    "                    CountE = CountE + tCount\n",
    "\n",
    "            elif Orien == 'C':\n",
    "                if Strt-20 < Posit and Strt+16 > Posit:\n",
    "                    SCount = SCount + tCount\n",
    "                elif Strt-220 < Posit and Strt+16 > Posit:\n",
    "                    CountS = CountS + tCount\n",
    "                elif End + 200 > Posit:\n",
    "                    CountE = CountE + tCount\n",
    "\n",
    "        else :\n",
    "            CCount = Count - SCount\n",
    "            output.write(tName+'\\t'+str(tStrt)+'\\t'+str(tEnd)+'\\t'+str(tLength)+'\\t'+str(Count)+'\\t'+str(SCount)+'\\t'+str(CCount)+'\\t'+str(CountS)+'\\t'+str(CountE)+'\\n')\n",
    "            Count =0\n",
    "            SCount = 0\n",
    "            CCount = 0\n",
    "            CountS = 0\n",
    "            CountE = 0\n",
    "            tName = Name\n",
    "            Count = Count + tCount\n",
    "            tStrt = Strt\n",
    "            tEnd = End\n",
    "            tLength = Length\n",
    "            if Orien == 'W':\n",
    "                if Strt+20 > Posit and Strt-16 < Posit:\n",
    "                    SCount = SCount + tCount\n",
    "                elif Strt+220 > Posit and Strt-16 < Posit:\n",
    "                    CountS = CountS + tCount\n",
    "                elif End - 200 < Posit:\n",
    "                    CountE = CountE + tCount\n",
    "\n",
    "            elif Orien == 'C':\n",
    "                if Strt-20 < Posit and Strt+16 > Posit:\n",
    "                    SCount = SCount + tCount\n",
    "                elif Strt-220 < Posit and Strt+16 > Posit:\n",
    "                    CountS = CountS + tCount\n",
    "                elif End + 200 > Posit:\n",
    "                    CountE = CountE + tCount\n",
    "output.close()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "820364.0\n"
     ]
    }
   ],
   "source": [
    "#Converts accumulated sorted data at the genome level to gene features\n",
    "\n",
    "file1 = open('C:/Zid/SolexaDataAnalysis/Anna_7_18_17/RNA3AlignSortCount.txt','r')\n",
    "output = open('C:/Users/Brian/OneDrive/ZidLab/Data/Solexa/RNA3Feat.txt', 'w')\n",
    "AlignCountTotal = 0.0\n",
    "\n",
    "for line in file1:\n",
    "    temp = line.rstrip()\n",
    "    temp1 = temp.split('\\t') \n",
    "    Chr = int(temp1[1])\n",
    "    Posit = int(temp1[3])\n",
    "    Posit = Posit + 1\n",
    "    Orien = temp1[0]\n",
    "    count = temp1[2]\n",
    "    AlignCount = 0\n",
    "    \n",
    "    #Seq = temp1[5]\n",
    "    if Chr == 1 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND1.txt','r')\n",
    "    elif Chr == 2 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND2.txt','r')\n",
    "    elif Chr == 3 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND3.txt','r')\n",
    "    elif Chr == 4 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND4.txt','r')\n",
    "    elif Chr == 5 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND5.txt','r')\n",
    "    elif Chr == 6 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND6.txt','r')\n",
    "    elif Chr == 7 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND7.txt','r')\n",
    "    elif Chr == 8 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND8.txt','r')\n",
    "    elif Chr == 9 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND9.txt','r')\n",
    "    elif Chr == 10 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND10.txt','r')\n",
    "    elif Chr == 11 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND11.txt','r')\n",
    "    elif Chr == 12 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND12.txt','r')\n",
    "    elif Chr == 13 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND13.txt','r')\n",
    "    elif Chr == 14 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND14.txt','r')\n",
    "    elif Chr == 15 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND15.txt','r')\n",
    "    elif Chr == 16 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND16.txt','r')\n",
    "    elif Chr == 17 :\n",
    "        file2 = open('C:/Zid/SolexaDataAnalysis/Chrom/ORF_ND17.txt','r')    \n",
    "    \n",
    "    for line2 in file2 :\n",
    "        temp3 = line2.rstrip()\n",
    "        temp2 = temp3.split('\\t')\n",
    "        Chr2 = int(temp2[1])\n",
    "        if Chr == Chr2: #checking if the same chromosome\n",
    "            if Orien == '-' :  # C orientation\n",
    "                name = temp2[0]\n",
    "                Orien2 = temp2[4]\n",
    "                if Orien2 == '-' :  #also C orientation\n",
    "                    tStrt2 = int(temp2[3])\n",
    "                    tEnd2 = int(temp2[2])\n",
    "                    Strt2 = tStrt2+16\n",
    "                    End2 = tEnd2+13\n",
    "                                    \n",
    "                    if End2 < Posit < Strt2 :\n",
    "                        AlignCount = AlignCount + 1\n",
    "                        output.write(name+'\\t'+str(tStrt2)+'\\t'+str(tEnd2)+'\\t'+str(Posit)+'\\t'+count+'\\n') #sense count\n",
    "                \n",
    "                  \n",
    "            else :          # W orientation\n",
    "                name = temp2[0]\n",
    "                Orien2 = temp2[4]\n",
    "                if Orien2 == '+' :  #also W orientation\n",
    "                    tStrt2 = int(temp2[2])\n",
    "                    tEnd2 = int(temp2[3])\n",
    "                    Strt2 = tStrt2-16\n",
    "                    End2 = tEnd2-13\n",
    "\n",
    "                    if Strt2 < Posit < End2 :\n",
    "                        AlignCount = AlignCount + 1\n",
    "                        output.write(name+'\\t'+str(tStrt2)+'\\t'+str(tEnd2)+'\\t'+str(Posit)+ '\\t'+count+'\\n') #sense count\n",
    "                        AlignCountTotal = AlignCountTotal + float(count)\n",
    "    file2.close()\n",
    "print AlignCountTotal\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv('C:/Users/Brian/OneDrive/ZidLab/Data/Solexa/GenesRibo.txt',sep='\\t',header=(0))\n",
    "\n",
    "df2 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260106_WT_ribo_rep1_counts.txt',sep='\\t',header=(0))\n",
    "df3 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260107_WT_ribo_rep2_counts.txt',sep='\\t',header=(0))\n",
    "df4 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260108_WT_ribo_rep3_counts.txt',sep='\\t',header=(0))\n",
    "df5 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260109_can1_ribo_rep1_counts.txt',sep='\\t',header=(0)) \n",
    "df6 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260110_can1_ribo_rep2_counts.txt',sep='\\t',header=(0))\n",
    "df7 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260111_can1_ribo_rep3_counts.txt',sep='\\t',header=(0))\n",
    "df8 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260112_WT_mRNA_rep1_counts.txt',sep='\\t',header=(0))\n",
    "df9 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260113_WT_mRNA_rep2_counts.txt',sep='\\t',header=(0)) \n",
    "df10 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260114_can1_mRNA_rep1_counts.txt',sep='\\t',header=(0))\n",
    "df11 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260115_can1_mRNA_rep2_counts.txt',sep='\\t',header=(0))\n",
    "df12 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260116_rpl22a_ribo_rep1_counts.txt',sep='\\t',header=(0))\n",
    "df13 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260117_rpl22a_ribo_rep2_counts.txt',sep='\\t',header=(0)) \n",
    "df14 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260118_tor1_ribo_rep1_counts.txt',sep='\\t',header=(0))\n",
    "df15 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260119_tor1_ribo_rep2_counts.txt',sep='\\t',header=(0))\n",
    "df16 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260120_rpl22a_mRNA_counts.txt',sep='\\t',header=(0))\n",
    "df17 = pd.read_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/GSE85198_RAW/GSM2260121_tor1_mRNA_counts.txt',sep='\\t',header=(0)) \n",
    "new_df = df.merge(df2, how='left', on='Name').merge(df3, how='left', on='Name').merge(df4, how='left', on='Name').merge(df5, how='left', on='Name').merge(df6, how='left', on='Name').merge(df7, how='left', on='Name').merge(df8, how='left', on='Name').merge(df9, how='left', on='Name').merge(df10, how='left', on='Name').merge(df11, how='left', on='Name').merge(df12, how='left', on='Name').merge(df13, how='left', on='Name').merge(df14, how='left', on='Name').merge(df15, how='left', on='Name').merge(df16, how='left', on='Name').merge(df17, how='left', on='Name')\n",
    "\n",
    "new_df.to_csv('D:/ZidLab/Solexa/OtherPeoples/YeastRplRibo/dataMerge.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2360930\n"
     ]
    }
   ],
   "source": [
    "#Combine AlignSortCount Files\n",
    "#Combine AlignCountSorts  10_8_2014\n",
    "file1 = open('D:/ZidData/OSheaLab/Solexa/GluTime2/Aligning/AlignedFiles/3P_S1_7p35AlignSortCount.txt') \n",
    "file2 = open('D:/ZidData/OSheaLab/Solexa/GluTime2/Aligning/AlignedFiles/3Pp35bAlignSortCount.txt') \n",
    "\n",
    "output = open('D:/ZidData/OSheaLab/Solexa/GluTime2/Aligning/AlignedFiles/3Pp35CombinedAlignSortCount.txt','w')\n",
    "\n",
    "count = 0\n",
    "temp = file1.readline()\n",
    "line1 = temp.rstrip() \n",
    "temp1 = line1.split('\\t')\n",
    "dct_lst = [{\"Orien\":temp1[0],\"Chr\":int(temp1[1]),\"Count\":int(temp1[2]),\"Posit\": temp1[3]}]\n",
    "for line in file1:\n",
    "    line1 = line.rstrip() \n",
    "    temp1 = line1.split('\\t')\n",
    "    temp_lst = [{\"Orien\":temp1[0],\"Chr\":int(temp1[1]),\"Count\":int(temp1[2]),\"Posit\": int(temp1[3])}]\n",
    "    dct_lst.extend(temp_lst)\n",
    "for line in file2:\n",
    "    line1 = line.rstrip() \n",
    "    temp1 = line1.split('\\t')\n",
    "    temp_lst = [{\"Orien\":temp1[0],\"Chr\":int(temp1[1]),\"Count\":int(temp1[2]),\"Posit\": int(temp1[3])}]\n",
    "    dct_lst.extend(temp_lst)\n",
    "    \n",
    "from operator import itemgetter\n",
    "dct_lst.sort(key=itemgetter(\"Posit\"))\n",
    "\n",
    "from operator import itemgetter\n",
    "dct_lst.sort(key=itemgetter(\"Chr\"))\n",
    "\n",
    "count2 = 0\n",
    "CountT = 0\n",
    "test = 0\n",
    "Length = len(dct_lst)\n",
    "for i in range(len(dct_lst)) :\n",
    "    Chr = (dct_lst[i][\"Chr\"])\n",
    "    Posit = (dct_lst[i][\"Posit\"])\n",
    "    Orien = (dct_lst[i][\"Orien\"])\n",
    "    CountTemp = (dct_lst[i][\"Count\"])\n",
    "    \n",
    "    if i == (Length-1):\n",
    "        if test == 1:\n",
    "            output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountT)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n')\n",
    "            count2 = CountT + count2\n",
    "        else:\n",
    "            output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountTemp)+'\\t'+str(Posit)+'\\t'+'\\n')\n",
    "            count2 = CountTemp + count2\n",
    "    else:\n",
    "        OrienT2 = (dct_lst[i+1][\"Orien\"])\n",
    "        PositT2 = (dct_lst[i+1][\"Posit\"])\n",
    "        CountT2 = (dct_lst[i+1][\"Count\"])\n",
    "        if Orien == OrienT2 :\n",
    "        \n",
    "            if Posit == PositT2 :\n",
    "                test = 1\n",
    "                CountT = CountT2 + CountTemp\n",
    "            else :\n",
    "                if test == 1:\n",
    "                    output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountT)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n')\n",
    "                    count2 = CountT + count2\n",
    "                    CountT = 0\n",
    "                else:\n",
    "                    output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountTemp)+'\\t'+str(Posit)+'\\t'+'\\n')\n",
    "                    count2 = CountTemp + count2\n",
    "                    CountT = 0\n",
    "                \n",
    "                test = 0\n",
    "        else :\n",
    "            \n",
    "            if test == 1:\n",
    "                output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountT)+'\\t'+str(Posit)+'\\t'+'\\n')#Seq+'\\n')\n",
    "                count2 = CountT + count2\n",
    "                CountT = 0\n",
    "            else:\n",
    "                output.write(Orien+'\\t'+str(Chr)+'\\t'+str(CountTemp)+'\\t'+str(Posit)+'\\t'+'\\n')\n",
    "                count2 = CountTemp + count2\n",
    "                CountT = 0\n",
    "            test = 0\n",
    "         \n",
    "output.close()  \n",
    "print count2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
